File: app/src/main/java/com/xeniac/warrantyroster/landingactivity/forgotpasswordfragment/ForgotPwSentFragment.java
package com.xeniac.warrantyroster.landingactivity.forgotpasswordfragment;

import android.app.Activity;
import android.content.Context;
import android.os.Bundle;
import android.os.CountDownTimer;
import android.os.Handler;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;

import com.apollographql.apollo.ApolloCall;
import com.apollographql.apollo.ApolloClient;
import com.apollographql.apollo.api.Response;
import com.apollographql.apollo.exception.ApolloException;
import com.google.android.material.snackbar.BaseTransientBottomBar;
import com.google.android.material.snackbar.Snackbar;
import com.xeniac.warrantyroster.Constants;
import com.xeniac.warrantyroster.NetworkHelper;
import com.xeniac.warrantyroster.R;
import com.xeniac.warrantyroster.databinding.FragmentForgotPwSentBinding;

import java.text.DecimalFormat;
import java.util.Arrays;

public class ForgotPwSentFragment extends Fragment {

    private FragmentForgotPwSentBinding forgotPwSentBinding;
    private View view;
    private Activity activity;
    private Context context;

    private String email;
    private CountDownTimer countDownTimer;

    public ForgotPwSentFragment() {
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        forgotPwSentBinding = FragmentForgotPwSentBinding.inflate(inflater, container, false);
        view = forgotPwSentBinding.getRoot();
        return view;
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        if (countDownTimer != null) {
            countDownTimer.cancel();
        }
        forgotPwSentBinding = null;
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        activity = getActivity();
        context = getContext();

        getEmailFromArgs();
        returnOnClick();
        resendOnClick();
    }

    private void getEmailFromArgs() {
        if (getArguments() != null) {
            ForgotPwSentFragmentArgs args = ForgotPwSentFragmentArgs.fromBundle(getArguments());
            email = args.getEmail();
        }
    }

    private void returnOnClick() {
        forgotPwSentBinding.btnForgotPwSentReturn.setOnClickListener(view ->
                activity.onBackPressed());
    }

    private void resendOnClick() {
        forgotPwSentBinding.btnForgotPwSentResend.setOnClickListener(view ->
                resendResetPasswordEmail());
    }

    private void resendResetPasswordEmail() {
        if (NetworkHelper.hasNetworkAccess(context)) {
            resendResetPasswordEmailMutation();
        } else {
            hideLoadingAnimation();
            Snackbar.make(view, context.getResources().getString(R.string.network_error_connection),
                    BaseTransientBottomBar.LENGTH_INDEFINITE)
                    .setAction(context.getResources().getString(R.string.network_error_retry),
                            v -> resendResetPasswordEmail()).show();
        }
    }

    private void resendResetPasswordEmailMutation() {
        showLoadingAnimation();

        ApolloClient apolloClient = ApolloClient.builder()
                .serverUrl(Constants.URL_GRAPHQL)
                .build();

        apolloClient.mutate(new SendResetPasswordEmailMutation(email))
                .enqueue(new ApolloCall.Callback<SendResetPasswordEmailMutation.Data>() {
                    @Override
                    public void onResponse(@NonNull Response<SendResetPasswordEmailMutation.Data> response) {
                        if (forgotPwSentBinding != null) {
                            activity.runOnUiThread(() -> {
                                hideLoadingAnimation();

                                if (!response.hasErrors()) {
                                    Log.i("resendResetPwEmail", "onResponse: " + response);
                                    countdown();
                                    forgotPwSentBinding.lavForgotPwSent.playAnimation();
                                } else {
                                    Log.e("resendResetPwEmail", "onResponse Errors: " + response.getErrors());
                                    if (Arrays.toString(response.getErrors().get(0).getCustomAttributes().values().toArray()).contains("code=500")) {
                                        Snackbar.make(view, context.getResources().getString(R.string.forgot_pw_sent_error_wait),
                                                BaseTransientBottomBar.LENGTH_LONG)
                                                .show();
                                    } else {
                                        Snackbar.make(view, context.getResources().getString(R.string.network_error_response),
                                                BaseTransientBottomBar.LENGTH_INDEFINITE)
                                                .setAction(context.getResources().getString(R.string.network_error_retry),
                                                        v -> resendResetPasswordEmail()).show();
                                    }
                                }
                            });
                        }
                    }

                    @Override
                    public void onFailure(@NonNull ApolloException e) {
                        Log.e("resendResetPwEmail", "onFailure: " + e.getMessage());
                        if (forgotPwSentBinding != null) {
                            activity.runOnUiThread(() -> {
                                hideLoadingAnimation();
                                Snackbar.make(view, context.getResources().getString(R.string.network_error_failure),
                                        BaseTransientBottomBar.LENGTH_LONG).show();
                            });
                        }
                    }
                });
    }

    private void showLoadingAnimation() {
        forgotPwSentBinding.btnForgotPwSentResend.setVisibility(View.GONE);
        forgotPwSentBinding.cpiForgotPwSent.setVisibility(View.VISIBLE);
    }

    private void hideLoadingAnimation() {
        forgotPwSentBinding.cpiForgotPwSent.setVisibility(View.GONE);
        forgotPwSentBinding.btnForgotPwSentResend.setVisibility(View.VISIBLE);
    }

    private void countdown() {
        forgotPwSentBinding.groupForgotPwSentResend.setVisibility(View.GONE);
        forgotPwSentBinding.groupForgotPwSentTimer.setVisibility(View.VISIBLE);

        DecimalFormat decimalFormat = new DecimalFormat("00");
        int startTime = 120 * 1000;

        countDownTimer = new CountDownTimer(startTime, 1) {
            @Override
            public void onTick(long millisUntilFinished) {
                long minutes = millisUntilFinished / 60000;
                long seconds = (millisUntilFinished / 1000) % 60;

                forgotPwSentBinding.tvForgotPwSentTimer.setText(String.format("(%s:%s)",
                        decimalFormat.format(minutes), decimalFormat.format(seconds)));
            }

            @Override
            public void onFinish() {
                new Handler().postDelayed(() -> {
                    forgotPwSentBinding.groupForgotPwSentTimer.setVisibility(View.GONE);
                    forgotPwSentBinding.groupForgotPwSentResend.setVisibility(View.VISIBLE);
                }, 500);
            }
        }.start();
    }
}

File: app/src/main/java/com/xeniac/warrantyrostermanager/landingactivity/forgotpasswordfragment/ForgotPwSentFragment.kt
package com.xeniac.warrantyrostermanager.landingactivity.forgotpasswordfragment

import android.os.Bundle
import android.os.CountDownTimer
import android.os.Handler
import android.os.Looper
import android.util.Log
import android.view.View
import android.view.View.GONE
import android.view.View.VISIBLE
import androidx.fragment.app.Fragment
import androidx.navigation.fragment.navArgs
import com.google.android.material.snackbar.BaseTransientBottomBar.LENGTH_INDEFINITE
import com.google.android.material.snackbar.BaseTransientBottomBar.LENGTH_LONG
import com.google.android.material.snackbar.Snackbar
import com.google.firebase.auth.FirebaseAuth
import com.xeniac.warrantyrostermanager.NetworkHelper
import com.xeniac.warrantyrostermanager.R
import com.xeniac.warrantyrostermanager.databinding.FragmentForgotPwSentBinding
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.tasks.await
import kotlinx.coroutines.withContext
import java.lang.Exception
import java.text.DecimalFormat

class ForgotPwSentFragment : Fragment(R.layout.fragment_forgot_pw_sent) {

    private var _binding: FragmentForgotPwSentBinding? = null
    private val binding get() = _binding!!

    private lateinit var firebaseAuth: FirebaseAuth
    private lateinit var email: String
    private lateinit var countDownTimer: CountDownTimer

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        _binding = FragmentForgotPwSentBinding.bind(view)
        firebaseAuth = FirebaseAuth.getInstance()

        getEmailFromArgs()
        returnOnClick()
        resendOnClick()
    }

    override fun onDestroyView() {
        super.onDestroyView()
        countDownTimer.cancel()
        _binding = null
    }

    private fun getEmailFromArgs() {
        val args: ForgotPwSentFragmentArgs by navArgs()
        email = args.email
    }

    private fun returnOnClick() {
        binding.btnForgotPwSentReturn.setOnClickListener {
            requireActivity().onBackPressed()
        }
    }

    private fun resendOnClick() {
        binding.btnForgotPwSentResend.setOnClickListener {
            resendResetPasswordEmail()
        }
    }

    private fun resendResetPasswordEmail() {
        if (NetworkHelper.hasNetworkAccess(requireContext())) {
            resendResetPasswordEmailAuth()
        } else {
            hideLoadingAnimation()
            Snackbar.make(
                binding.root,
                requireContext().getString(R.string.network_error_connection),
                LENGTH_INDEFINITE
            ).apply {
                setAction(requireContext().getString(R.string.network_error_retry)) { resendResetPasswordEmail() }
                show()
            }
        }
    }

    private fun resendResetPasswordEmailAuth() {
        showLoadingAnimation()
        CoroutineScope(Dispatchers.IO).launch {
            try {
                firebaseAuth.sendPasswordResetEmail(email).await()
                Log.i(
                    "resendResetPasswordEmail",
                    "Reset password email successfully sent to ${email}."
                )
                withContext(Dispatchers.Main) {
                    hideLoadingAnimation()
                    countdown()
                    binding.lavForgotPwSent.playAnimation()
                }
            } catch (e: Exception) {
                Log.e("resendResetPasswordEmail", "Exception: ${e.message}")
                withContext(Dispatchers.Main) {
                    hideLoadingAnimation()
                    Snackbar.make(
                        binding.root,
                        requireContext().getString(R.string.network_error_failure),
                        LENGTH_LONG
                    ).show()
                }
            }
        }
    }

    private fun showLoadingAnimation() {
        binding.btnForgotPwSentResend.visibility = GONE
        binding.cpiForgotPwSent.visibility = VISIBLE
    }

    private fun hideLoadingAnimation() {
        binding.btnForgotPwSentResend.visibility = VISIBLE
        binding.cpiForgotPwSent.visibility = GONE
    }

    private fun countdown() {
        binding.groupForgotPwSentResend.visibility = GONE
        binding.groupForgotPwSentTimer.visibility = VISIBLE

        val decimalFormat = DecimalFormat("00")
        val startTime = 120 * 1000

        object : CountDownTimer(startTime.toLong(), 1) {
            override fun onTick(millisUntilFinished: Long) {
                val minutes = millisUntilFinished / 60000
                val seconds = (millisUntilFinished / 1000) % 60

                binding.tvForgotPwSentTimer.text =
                    "(${decimalFormat.format(minutes)}:${decimalFormat.format(seconds)})"
            }

            override fun onFinish() {
                Handler(Looper.getMainLooper()).postDelayed({
                    binding.groupForgotPwSentTimer.visibility = GONE
                    binding.groupForgotPwSentResend.visibility = VISIBLE
                }, 500)
            }
        }.start()
    }
}